### PR TITLE
Update cinder.conf config template with Liberty

### PR DESCRIPTION
**Pull request by mjura**
Updating cinder.conf template config with Liberty

### CODE REVIEW

**Comment by vuntz** [comment]
@mjura can you confirm that cinder_emc_config.xml.erb and cinder_eternus_dx_config.xml.erb are still correct?

**Comment by vuntz** [comment]
@mjura I see a huge diff with the sample config from liberty (I can provide it if wanted). Is this ready?

**Comment by mjura** [comment]
@vuntz : We've decided to rework this little bit and split it in different commits. Sample which I'm using for was generated by tox https://github.com/openstack/cinder/blob/stable/liberty/etc/cinder/README-cinder.conf.sample. In config-refences there is also sample but has some deprecated options.

**Comment by vuntz** [comment]
@mjura I h

**Comment by mjura** [comment]
Cinder.conf.erb was synchronized with Liberty config sample. Because multi backend support I have to organize backend drivers section. I've tested it and works fine.

**Comment by mjura** [comment]
@vuntz : Fujitsu Eternus driver was removed from Liberty, I have created another card to re-enable this driver for SOC6 https://trello.com/c/H9qIkgWh/794-re-establish-fujitsu-eternus-dx-driver-in-cinder-liberty

**Comment by mjura** [comment]
I've updated cinder_emc_config.xml.erb template, but for cinder_eternus_dx_config.xml.erb this will be done in other history card.

**Comment by mjura** [comment]
@aplanas : Thank you, I fixed this

**Comment by aplanas** [comment]
+1

**Comment by toabctl** [comment]
@mjura needs a rebase...

**Comment by mjura** [comment]
Rebased

**Comment by toabctl** [comment]
+1

**Comment by aplanas** [comment]
+1

**Comment by dirkmueller** [comment]
-1, the migration also needs to update the volumes array for each backend that uses the emc backend

**Comment by vuntz** [comment]
@mjura I still have a huge diff with the sample config from liberty cinder here. Maybe my sample is wrong, though. Where did you pick yours / how did you generate it?

**Comment by mjura** [comment]
@vuntz : It is cinder.conf.sample generated by `tox -egenconf` command, but I adopt it to multi backend support.

**Comment by toabctl** [comment]
> @vuntz : It is cinder.conf.sample generated by tox -egenconf command, but I adopt it to multi backend support.

@mjura you should use the sample config from the Liberty package. The generated configs can be different depending on the versions of i.e. oslo.* .

**Comment by mjura** [comment]
@toabctl : thank you for advice, I've checked latest cinder.conf from openstack-cinder package and it is too small for our needs.

**Comment by mjura** [comment]
I've fixed migration script and rebased this change

**Comment by vuntz** [comment]
> I've checked latest cinder.conf from openstack-cinder package and it is too small for our needs.

It should have been fixed yesterday; which version did you check, just to be sure?

**Comment by mjura** [comment]
@vuntz : I was checking package openstack-cinder-7.0.1~a0~dev35-1.1.noarch.rpm.
But anyway I think that, template submited to this PR is enough good to use it. It was generated by `tox -egenconfig`

**Comment by vuntz** [comment]
> But anyway I think that, template submited to this PR is enough good to use it. It was generated by tox -egenconfig

I'm not saying it's not correct, but I'm saying it'll be painful because it's totally different from our sample (and even from the one I generated with tox). So there'll be another huge PR to rebase the config. I don't think we want another huge PR for this.

**Comment by mjura** [comment]
@vuntz: As I said this is final change. This config sample was generated with `tox -egenconfig` and adopted for multi-backend support.

**Comment by mjura** [comment]
Rebased

**Comment by dirkmueller** [comment]
goal was not to use tox but to use the packaged versions

**Comment by mjura** [comment]
> goal was not to use tox but to use the packaged versions

Heh, there is no where written that it has to be used from package. In the past we used samples from config-references or from github repo. As I said in comment https://github.com/crowbar/crowbar-openstack/pull/95#issuecomment-159911642, when I was checking package sample it was broken. But espacially for you I will check if it was fixed in latest version of our packages. Anyway samples in packages are using also tox with scripts on top and maybe different versions python packages.

**Comment by mjura** [comment]
It was rebased

**Comment by mjura** [comment]
@dirkmueller : I've checked `cinder.conf` file in package `openstack-cinder-7.0.1~a0~dev35-6.1.noarch.rpm` from D:C:6:S, it has 828 lines in comparison to 3255 lines file generated by `tox -egenconfig`. Please let me what file I have to use to update this template. I just don't care about this, but I want to have clear requirements `Please use this file to update`

**Comment by vuntz** [comment]
For the record: package with correct sample config is in OBS/C:O:L:S (still in staging as gating broken).

**Comment by mjura** [comment]
Uff, I think that I did it third time :) this time it was synchronized with `cinder.conf` from `openstack-cinder-7.0.1~a0~dev42-38.2.noarch.rpm` package from OBS/C:O:L:S.

Please review

**Comment by mjura** [comment]
Rebased

**Comment by mjura** [comment]
Was tested with LVM and Ceph backend

File: chef/data_bags/crowbar/migrate/cinder/044_update_emc_backend.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+def upgrade(ta, td, a, d)
+  a["volume_defaults"]["ecom_server_portgroup_1"] = ta["volume_defaults"]["ecom_server_portgroup_1"]
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/cinder/044_update_emc_backend.rb]: thats should be 

```ruby
a["volume_defaults"]["emc"]
```

File: chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb [hunk 2]
```diff
@@ -1,9 +1,14 @@
...
+<?xml version="1.0" encoding="UTF-8" ?>
 <EMC>
-    <StorageType><%= @emc_params["emc_storage_type"] %></StorageType>
-    <MaskingView><%= @emc_params["masking_view"] %></MaskingView>
     <EcomServerIp><%= @emc_params["ecom_server_ip"] %></EcomServerIp>
     <EcomServerPort><%= @emc_params["ecom_server_port"] %></EcomServerPort>
     <EcomUserName><%= @emc_params["ecom_server_username"] %></EcomUserName>
     <EcomPassword><%= @emc_params["ecom_server_password"] %></EcomPassword>
+    <PortGroups>
+      <PortGroup><%= @emc_params["ecom_server_portgroup_1"] %></PortGroup>
```
Comment (dirkmueller) [inline chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb]: are you sure that this is only two portgroups and exactly two portgroups? that looks broken

Comment (mjura) [inline chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb]: Yes, you are right I'm not sure about this, but I can't check this right now. I was trying to get access to emc storage controller. Testing EMC support will covered by another trello card.

File: chef/data_bags/crowbar/migrate/cinder/044_update_emc_backend.rb (hunk 1 above)
Comment (mjura) [inline chef/data_bags/crowbar/migrate/cinder/044_update_emc_backend.rb]: I fixed it

**Comment by mjura** [comment]
It was rebased

**Comment by mjura** [comment]
It was rebased

**Comment by toabctl** [comment]
@mjura I still get a huge diff compared to the config from the package...

**Comment by aplanas** [comment]
For me, if / when gating is green: +1

**Comment by mjura** [comment]
It was rebased,

> @mjura I still get a huge diff compared to the config from the package...

@toabctl : We discussed this couple of times also during manila multi backend template update. I used proper `cinder.conf.sample` but I moved specific hardware backend options to multi backend section. This was done for Cloud 5 and Cloud 4 as well and @vuntz told me that we have to stay with this.

**Comment by dirkmueller** [comment]
do not merge before M8

**Comment by mjura** [comment]
It was only rebased

**Comment by mjura** [comment]
It was rebased

**Comment by mjura** [comment]
Rebased

File: chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb (hunk 2 above)
Comment (dirkmueller) [inline chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb]: then why not leave it out? I don't think this is a proper way of doing it.

**Comment by mjura** [comment]
@toabctl: thank you, I wanted to proceed with this change too and remove `do not merge yet` label

**Comment by toabctl** [comment]
@mjura needs rebase.

File: chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb (hunk 2 above)
Comment (mjura) [inline chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb]: It was done according to documentation http://docs.openstack.org/liberty/config-reference/content/emc-vmax-driver.html

**Comment by mjura** [comment]
@toabctl: it was rebased

**Comment by toabctl** [comment]
Not sure howto continue here. So my current config (extracted from Devel:Clloud:6:Staging/openstack-cinder) is here: http://paste.openstack.org/show/484380/  . The diff to the cinder.conf.erb is here: http://paste.openstack.org/show/484381/ . Looks still huge...

**Comment by mjura** [comment]
@toabctl : Please make diff once again, I sync config with cinder.conf file from openstack-cinder-7.0.2-a0-dev9-84.1.noarch.rpm package after fix for sorting options

**Comment by toabctl** [comment]
@mjura Diff to config from package looks good now. thanks!

**Comment by toabctl** [comment]
In ```[keystone_authtoken]``` , the parameter ```region_name``` must be set to ```region_name = <%= @keystone_settings['endpoint_region'] %>```

File: chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb (hunk 2 above)
Comment (toabctl) [inline chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb]: I'm also for leaving the portgroup out for now. Sounds wrong to define only 2...

**Comment by mjura** [comment]
@toabctl: Thank you for comment, I've updated [keystone_authtoken] section and `region_name`

**Comment by toabctl** [comment]
@mjura what about the portgroup?

**Comment by dirkmueller** [comment]
it looks like you're removing eternus support: 

```diff
-<% if volume['backend_driver'] == 'eternus' -%>
-#
-# Options defined in cinder.volume.drivers.fujitsu_eternus_dx_common
-#
+# Whether to unmap volume before deletion. (boolean value)
+#sio_unmap_volume_before_deletion = false
 
-# The configuration file for the Cinder SMI-S driver (string
-# value)
-cinder_smis_config_file=/etc/cinder/cinder_eternus_dx_config-<%= backend_id %>.xml

```

**Comment by mjura** [comment]
@dirkmueller : I've added once again `fujitsu_eternus_dx_common` configuration to cinder.conf. I was thinking that it will be part of the card about reintroducing Fujitsu Eternus driver.

@dirkmueller : We talked with @toabctl about Port Group in EMC driver and these parameters are obligatory to leave. My proposal is to leave it as it is and it will validated in follow up card about testing EMC driver. Card was already created.

**Comment by dirkmueller** [comment]
Looks like there is a mismatch:

```ruby
<% if volume['backend_driver'] == 'zfssa' -%>
volume_driver = cinder.volume.drivers.zfssaiscsi.ZFSSAISCSIDriver
<% end -%>
...
<% if volume['backend_driver'] == 'zfssaiscsi' -%>
```

**Comment by mjura** [comment]
Ok, I fixed `zfssa` driver part

**Comment by dirkmueller** [comment]
@mjura who is "we" ? What do you do now if you have a EMC with only one portgroup? you require two. it seems easier imho to not require portgroups (what was suggested already) or to use array pools instead. As it is now it is imho wrong.

**Comment by dirkmueller** [comment]
also, the LVM driver got renamed to 

```ruby
cinder.volume.drivers.lvm.LVMVolumeDriver
```

**Comment by mjura** [comment]
@dirkmueller : We updated LVM driver and portgroup for EMC will be created as array now, I have to only test it. Do you have more comments besides this ?

**Comment by mjura** [comment]
Ok, it was tested. I've added array to EMC Port Group option :) as you wanted @dirkmueller

**Comment by toabctl** [comment]
+1

**Comment by dirkmueller** [comment]
@mjura volume_driver entry for EMC is still wrong, you're referencing the dropped driver..

**Comment by mjura** [comment]
@dirkmueller : Thank you for pointing this, I've setup 
```
<% if volume['backend_driver'] == 'emc' -%>
volume_driver = cinder.volume.drivers.emc.emc_vmax_iscsi.EMCVMAXISCSIDriver
<% end -%>
```
Btw. I think that it could be done during EMC storage validation card.

**Comment by dirkmueller** [comment]
@mjura I'm fine with leaving out the EMC porting that you did, as written before already. 

can you please readd the eternus config reference and fix the LVM driver?

**Comment by mjura** [comment]
@dirkmueller : LVM driver is already configured as
```
volume_driver = cinder.volume.drivers.lvm.LVMVolumeDriver
```
And Fujitsu Eternus driver is as follows
```
<% if volume['backend_driver'] == 'eternus' -%>
  <% if volume['eternus']['protocol'] == "iscsi" -%>
volume_driver = cinder.volume.drivers.fujitsu_eternus_dx_iscsi.FJDXISCSIDriver
  <% else -%>
volume_driver = cinder.volume.drivers.fujitsu_eternus_dx_fc.FJDXFCDriver
  <% end -%>
<% end -%>
```
...
```
<% if volume['backend_driver'] == 'eternus' -%>
#
# Options defined in cinder.volume.drivers.fujitsu_eternus_dx_common
#

# The configuration file for the Cinder SMI-S driver (string
# value)
cinder_smis_config_file=/etc/cinder/cinder_eternus_dx_config-<%= backend_id %>.xml


<% end -%>
```

**Comment by dirkmueller** [comment]
@mjura sorry, apparently the last force-pull didn't work. There is still an issue with the eternus volume driver definition. it should be:

```bash
volume_driver=cinder.volume.drivers.fujitsu.eternus_dx_fc.FJDXFCDriver
```
and
```bash
volume_driver=cinder.volume.drivers.fujitsu.eternus_dx_iscsi.FJDXISCSIDriver
```

File: crowbar_framework/config/locales/cinder/en.yml [hunk 3]
```diff
@@ -64,11 +65,13 @@ en:
...
             emc:
               ecom_server_ip: 'IP address of the ECOM server'
-              ecom_server_password: 'Password for accessing the ECOM server'
               ecom_server_port: 'Port of the ECOM server'
               ecom_server_username: 'Username for accessing the ECOM server'
-              emc_storage_type: 'Thin pool where user wants to create volume from'
-              masking_view: 'Unisphere for VMAX Masking View'
+              ecom_server_password: 'Password for accessing the ECOM server'
+              ecom_server_portgroups: 'VMAX port groups to expose volumes managed by this backend'
+              ecom_server_array: 'Array unique VMAX array serial number'
```
Comment (dirkmueller) [inline crowbar_framework/config/locales/cinder/en.yml]: "Serial number of the VMAX  Array"

File: crowbar_framework/app/views/barclamp/cinder/_edit_attributes.html.haml [hunk 4]
```diff
@@ -86,12 +86,15 @@
...
               = string_field %w(volumes {{@index}} emc ecom_server_ip)
-              = string_field %w(volumes {{@index}} emc masking_view)
               = integer_field %w(volumes {{@index}} emc ecom_server_port)
               = string_field %w(volumes {{@index}} emc ecom_server_username)
               = password_field %w(volumes {{@index}} emc ecom_server_password)
-              = string_field %w(volumes {{@index}} emc emc_storage_type)
-
+              = array_string_field %w(volumes {{@index}} emc ecom_server_portgroups), :only_comma => true
+              %span.help-block
+                = t('.volumes.emc_server_portgroups')
```
Comment (dirkmueller) [inline crowbar_framework/app/views/barclamp/cinder/_edit_attributes.html.haml]: can you call it something with "_hint" ? that is the usual convention for help texts..

**Comment by mjura** [comment]
I've updated eternus driver name, and fix _hint for EMC bit

File: chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb [hunk 5]
```diff
@@ -1,9 +1,15 @@
...
+<?xml version="1.0" encoding="UTF-8" ?>
 <EMC>
-    <StorageType><%= @emc_params["emc_storage_type"] %></StorageType>
-    <MaskingView><%= @emc_params["masking_view"] %></MaskingView>
     <EcomServerIp><%= @emc_params["ecom_server_ip"] %></EcomServerIp>
     <EcomServerPort><%= @emc_params["ecom_server_port"] %></EcomServerPort>
     <EcomUserName><%= @emc_params["ecom_server_username"] %></EcomUserName>
     <EcomPassword><%= @emc_params["ecom_server_password"] %></EcomPassword>
+    <PortGroups>
+<% @emc_params["ecom_server_portgroups"].each_with_index do |portgroup, index| -%>
```
Comment (dirkmueller) [inline chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb]: you're not using the index so I guess .each would be enough..

**Comment by dirkmueller** [comment]
+1

File: chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb (hunk 5 above)
Comment (mjura) [inline chef/cookbooks/cinder/templates/default/cinder_emc_config.xml.erb]: Yes, it is tru I wanted to use also index, but it works fine without this. I change to `.each`

**Pull request merged**